ArduPlane/mode.h

```diff
@@ -0,0 +1,475 @@
+#pragma once
+
+#include <stdint.h>
+
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#include <AP_Math/AP_Math.h>
+
+#include "defines.h"
+
+#define MODE_NEXT_HEADING_UNKNOWN   99999.0f    // used to indicate to set_desired_location method that next leg's heading is unknown
+
+// pre-define ModeRTL so Auto can appear higher in this file
+class ModeRTL;
+
+class Mode
+{
+public:
+
+    // Auto Pilot modes
+    // ----------------
+    enum Number {
+        MANUAL        = 0,
+        CIRCLE        = 1,
+        STABILIZE     = 2,
+        TRAINING      = 3,
+        ACRO          = 4,
+        FLY_BY_WIRE_A = 5,
+        FLY_BY_WIRE_B = 6,
+        CRUISE        = 7,
+        AUTOTUNE      = 8,
+        AUTO          = 10,
+        RTL           = 11,
+        LOITER        = 12,
+        AVOID_ADSB    = 14,
+        GUIDED        = 15,
+        INITIALISING  = 16,
+        QSTABILIZE    = 17,
+        QHOVER        = 18,
+        QLOITER       = 19,
+        QLAND         = 20,
+        QRTL          = 21
+    };
+
+    // Constructor
+    Mode();
+
+    // enter this mode, returns false if we failed to enter
+    bool enter();
+
+    // perform any cleanups required:
+    void exit();
+
+    // returns a unique number specific to this mode
+    virtual uint32_t mode_number() const = 0;
+
+    // returns short text name (up to 4 bytes)
+    virtual const char *name4() const = 0;
+
+    //
+    // methods that sub classes should override to affect movement of the vehicle in this mode
+    //
+
+    // convert user input to targets, implement high level control for this mode
+    virtual void update() = 0;
+
+    //
+    // attributes for mavlink system status reporting
+    //
+
+    // returns true if any RC input is used
+    virtual bool has_manual_input() const { return false; }
+
+    // true if heading is controlled
+    virtual bool attitude_stabilized() const { return true; }
+
+    //
+    // navigation methods
+    //
+
+    // return distance (in meters) to destination
+    virtual float get_distance_to_destination() const { return 0.0f; }
+
+protected:
+
+    // subclasses override this to perform checks before entering the mode
+    virtual bool _enter() { return true; }
+
+    // subclasses override this to perform any required cleanup when exiting the mode
+    virtual void _exit() { return; }
+};
+
+
+class ModeAcro : public Mode
+{
+public:
+    uint32_t mode_number() const override { return Number::ACRO; }
+    const char *name4() const override { return "ACRO"; }
+
+    // methods that affect movement of the vehicle in this mode
+    void update() override { }
```

Comment:
This empty implementation should probably go in the base class.

Reply:
so, remove the pure virtual?

Reply:
Let's save it for now